README.md









Consistent Trees Merger Tree Code

Most code: Copyright (C)2011-2015 Peter Behroozi

License: GNU GPLv3

Science/Documentation Paper: <http://arxiv.org/abs/1110.4370>

## Contents
* [Compiling](#markdown-header-compiling)
* [Running](#markdown-header-running)
    1. [Quick Start for Rockstar Users](#markdown-header-quick-start-for-rockstar-users)
    2. [Guide for Other Halo Finders](#markdown-header-guide-for-other-halo-finders)
    3. [Including Extra Parameters](#markdown-header-including-extra-parameters)
    4. [Using Binary Input and Intermediate Formats](#markdown-header-using-binary-input-and-intermediate-formats)
    5. [Configuration Options](#markdown-header-configuration-options)
    6. [Running the Code](#markdown-header-running-the-code)
    7. [Troubleshooting](#markdown-header-troubleshooting)
* [Using the Trees](#markdown-header-using-the-trees)
    1. [Reading the Tree Files](#markdown-header-reading-the-tree-files)
    2. [Sussing Format](#markdown-header-sussing-format)
    3. [Halo Statistics](#markdown-header-halo-statistics)



## Compiling ##

If you use the GNU C compiler version 4.0 or above on a 64-bit machine,
compiling should be as simple as typing `make` at the command prompt.

The tree code does not support compiling on 32-bit machines and has not been
tested with other compilers.  Additionally, the tree code does not support
non-Unix environments. (Mac OS X is fine; Windows is not).  OpenMP is
now required to compile.

The tree code can use up to 32 cores, but is not completely parallelized.
In its current version, it can be reasonably used with up to 4096^3-particle
simulations.

## Running ##

1. ### Quick Start for Rockstar Users ###

    If you already have halo catalogs generated using the Rockstar Halo
    Finder (<https://bitbucket.org/gfcstanford/rockstar>), running is especially easy.
    
    The Rockstar output files (`out_*.list`) are already in the correct input
    format.  To generate all the output directories and the config file,
    run the following command:
        
        prompt> perl /path/to/rockstar/scripts/gen_merger_cfg.pl <rockstar.cfg>
    
    where `<rockstar.cfg>` refers to your rockstar configuration file.
    Then, follow the instructions the script gives you to run the merger tree
    code.
    
    If you run into problems with this, make sure that you are running the
    script from the same working directory as you ran Rockstar.  If you
    still have problems, see below for manual instructions.
    
2. ### Guide for Other Halo Finders ###

    
    You will need to generate several files in order to run the merger tree
    code.  First, you will need to convert your halo catalogs into the correct
    input format.  The merger tree code currently accepts input only in
    ASCII files with one halo per line.  The columns should be in the
    following order:
        
        #ID DescID Mass Vmax Vrms Radius Rs Np X Y Z VX VY VZ JX JY JZ Spin
    
    The columns should contain:
    
    +  `ID`: the halo ID, which must be unique across a single snapshot and must
    be at least 0.
    +  `DescID`: the halo's descendant id at the next timestep.  If no
    descendant exists, this should be `-1`.  At least 100 halos should
    have descendants (identified by particle-based methods) in order for
    the consistency checks to work properly.
    +  `Mass`: the halo mass.  This does *not* have to be Mvir, but it must
    correspond to the mass with the radius in the `Radius` column.  The *units* for this must be
    Msun/h.
    +  `Vmax`: the maximum circular velocity, in units of km/s (physical, not comoving).
    +  `Vrms`: the velocity dispersion in units of
    km/s (physical, not comoving); may be set to 0 if not available.
    +  `Radius`: the radius at which the mass is calculated in column 3.  The
    *units* must be in kpc / h.  (*Note* that this is different from the
    position units!)
    +  `Rs`: the scale radius of the halo, in 
    units of kpc / h; may be set to 0 if not available.
    +  `Np`: number of particles in the halo.
    +  `X`/`Y`/`Z`: 3D position of the halo, in units of Mpc / h.  (*Note* that this
    is different from the radius units!)
    +  `VX`/`VY`/`VZ`: 3D velocity of the halo, in units of km/s (physical, not comoving).
    +  `JX`/`JY`/`JZ`: 3D angular momentum of the halo,
    in units of (Msun/h) * (Mpc/h) * km/s (physical, not comoving); may be set to 0 if not available.
    +  `Spin`: Dimensionless spin parameter; may be set to 0 if not available.
    
    
    Note that non-numbers (like `NaN` and `Inf`) are *not acceptable* as inputs
    and may cause either the halo or the entire file to be rejected. 
    
    Here's a sample input from a scale factor of 0.075:
    
        
        #ID DescID Mvir Vmax Vrms Rvir Rs Np X Y Z VX VY VZ JX JY JZ Spin
        165 305 6.1824e+10 210.91 213.66 106.167 13.025 49 63.97294 100.14336 36.89451 78.44 82.65 117.79 3.852e+09 9.976e+08 -9.825e+09 0.14646
        166 307 9.3673e+09 100.62 108.34 56.599 14.260 34 28.23423 88.26684 27.98982 -42.16 112.89 61.36 -2.751e+08 -8.645e+07 -1.513e+08 0.10760
        167 309 1.4988e+10 119.98 127.34 66.199 14.129 22 30.63638 101.30433 25.59135 64.70 106.94 115.71 -5.153e+08 8.236e+08 -9.118e+07 0.16328
        168 310 2.4355e+10 135.83 144.66 77.827 24.273 22 98.15601 12.42515 30.58211 58.75 31.41 66.24 -9.074e+08 6.611e+08 -9.837e+08 0.05737
        3 -1 3.7469e+09 73.56 139.43 41.702 11.390 36 12.10802 0.52037 157.83209 -57.47 -11.73 -119.23 -1.918e+08 -4.043e+07 5.996e+08 1.72959
    
    
    Once you have created a conversion script, you should save each timestep
    in a file called "`out_XYZ.list`," where `XYZ` is the timestep number.
    For example, if you had three timesteps, you would save the first in
    "`out_0.list`," the second in "`out_1.list`," and the third in "`out_2.list`."
    
    Finally, you must create a file listing the number and the scale factor
    for each snapshot, one combination per line, in order of increasing
    scale factor.  For example:
        
        0 0.100
        1 0.120
        2 0.140
        ...
        98 0.990
        99 1.000
    
    You should save this in a file called "`DescScales.txt`" or similar.
    
3. ### Including Extra Parameters ###

    If your halo finder calculates extra parameters that you wish to propagate through the
    merger trees, you'll have to add a few options to the config file:
        
        EXTRA_PARAMS = <number of extra parameters>
    
    This number of parameters must correspond exactly to additional
    columns in the `out_*.list` files.  To keep things straight, you should
    also specify
        
        EXTRA_PARAM_LABELS = "label of each extra column"
        EXTRA_PARAM_DESCRIPTIONS = "#Desc of column 1\n#Desc of column 2..."
    
    to make sure that users of your merger trees understand what the
    extra columns mean.
    
    Finally, you may optionally specify
        
        EXTRA_PARAM_INTERPOLATIONS = "..."
    
    This should be a list of characters, one per extra parameter, to specify
    whether the extra parameter should be linearly interpolated ("`l`"),
    or whether the square or the cube of the parameter should be linearly
    interpolated ("`s`" or "`c`," respectively).  Most parameters should be
    linearly interpolated; however, parameters which are related by non-linear
    power laws (e.g., halo mass and halo radius) should make use of the
    alternate methods (linear for the halo mass and linear-cubic for the
    radius, in this example).  No commas or other separators should be used;
    for example, with five extra parameters, the value of this config
    option defaults to "`lllll`."
    
4. ### Using Binary Input and Intermediate Formats ###

    Consistent Trees can run faster if binary input formats are used (note that output formats
    will still be `ASCII`-formatted).  To use, generate `ASCII` inputs as above
    and run
        
        prompt>  /path/to/consistent/trees/convert_format merger_tree.cfg out_0.list ...
    
    This will generate binary versions of the `ASCII` inputs, appending `.bin` to each filename.
    The `merger_tree.cfg` is required in order to interpret the input files, as the input files
    may have extra columns (see `EXTRA_PARAMS` above).  In order to use the binary files
    with Consistent Trees, you will then have to specify "`INPUT_FORMAT = BINARY`" in the
    config file.  The `convert_format` command automatically recognizes whether the input
    is in binary or `ASCII` format, so the same command can be used to convert files back into
    `ASCII` formats:
    
        
        prompt>  /path/to/consistent/trees/convert_format merger_tree.cfg out_0.list.bin ...
    
    
    This will generate `ASCII` files with "`.txt`" appended to the input filenames. Note that
    specifying `merger_tree.cfg` is again required in order to have the correct columns in the
    output file.
    
5. ### Configuration Options ###

    It's best to copy one of the example scripts already provided 
    (e.g., `bolshoi.cfg` or `consuelo.cfg`) and to modify it as necessary.
    
    Config options you need to modify:
        
        Om = 0.27 # Omega Matter
        Ol = 0.73 # Omega Lambda
        h0 = 0.70 # h0
        
        BOX_WIDTH = 250 # The size of the simulation in Mpc/h
        BOX_DIVISIONS = 5 # For large sims, putting all the halos in one tree
        # file gets too unwieldy.  If BOX_DIVISIONS is larger than 1, then
        # the trees will be split into BOX_DIVISIONS^3 files, each
        # containing a cubic subregion of the box.
        
        SCALEFILE = "/path/to/DescScales.txt"
        INBASE = "/path/to/halo/catalog/directory" #Where out_*.list are. 
        OUTBASE = "/path/for/intermediate/steps/and/logfiles"
        TREE_OUTBASE = "/path/for/tree/files"
        HLIST_OUTBASE = "/path/for/output/halo/catalogs"
        
        #The last three directories are for storing the outputs of the tree code.
        #The "OUTBASE" will contain intermediate calculations, statistics, and
        #detailed logfiles about which halos are created or destroyed (and why).
        #These files can occupy a significant amount of space, on the order of
        #5x the size of the original input files.
        
        MASS_RES_OK=1e11 #Halo mass at which there are at least 1000 particles
        #in the halo
        
        #These options set the number of timesteps over which the halo has to
        # appear in order to be considered "valid."    
        MIN_TIMESTEPS_TRACKED = 5
        MIN_TIMESTEPS_SUB_TRACKED = 10
        MIN_TIMESTEPS_SUB_MMP_TRACKED = 10
        #Generally, MIN_TIMESTEPS_TRACKED should be set to 1/30 of the number of
        # timesteps, and MIN_TIMESTEPS_SUB_* should be set to 1/15 the number of
        # timesteps; these limits are a good balance between catching spurious
        # halos but also allowing for halos which merge soon after they appear.
        
        #Option for applying fix for Rockstar spins (only for versions prior
        #to Rockstar v0.99.9-RC3).  Set to 0-indexed column of T/|U| in EXTRA_PARAMS
        #E.g., if EXTRA_PARAM_LABELS is "Spin_bullock M200c T/|U|", then
        #FIX_ROCKSTAR_SPINS would be set to 2.
        FIX_ROCKSTAR_SPINS = -1
        
        #Option to limit memory usage; set to 1 if calculating merger tree for a large
        #(>2048^3 particles) box.  This turns off gzipping intermediate files and also
        #limits some parallelization to avoid having too many snapshots in memory.
        LIMITED_MEMORY = 0
    
    
    The other configuration options should generally be left alone unless you are
    doing development work on the tree code.
    
    
6. ### Running the Code ###

    
    Once you've compiled the code and created the config file, you can
    run the tree code.  If you have a periodic box, run
        
        cd /path/to/consistent_trees
        perl do_merger_tree.pl myconfig.cfg
    
    If you have a non-periodic box, then you should run instead:
        
        perl do_merger_tree_np.pl myconfig.cfg
    
    This will generate the merger trees.  To generate halo catalogs from the
    merger trees (which include properties like Vmax at accretion, etc.),
    then once the above command finishes, you should run
        
        perl halo_trees_to_catalog.pl myconfig.cfg
    
    (Again from the same directory as before).
    
7. ### Troubleshooting ###

    The MOST COMMON failure for the tree code is the following error:
        
        Error: too few halos at scale factor 0.XYZ to calculate consistency metric.
        Please remove this and all earlier timesteps from the scale file and rerun.
    
    If you get this, it means that there are too few halos at high redshifts
    to reliably calculate whether halos are consistent between timesteps or
    not.  To fix this problem, edit your DescScales.txt file and remove any
    timesteps which are at or before the scale factor mentioned, and then
    rerun the tree code.  This should solve the problem in 99\% of all cases.
    
## Using the Trees ##

1. ### Reading the Tree Files ###

    Included with the tree code is some example code for reading in the
    tree files produced.  (Not the halo catalogs, however).  This code
    is located here:
        
        /path/to/consistent_trees/read_tree
    
    The included `Makefile` shows how to compile the code into an executable,
    and the included `example.c` file shows how to call the reading procedure.
    
    Once the tree is loaded, you can access the halo tree from the
    `halo_tree` structure, and the list of all halos from the `all_halos`
    structure.  The definitions for these structures are in `read_tree.h`.
    
2. ### Sussing Format ###

    Merger trees in the Sussing Merger Trees format can be generated by
    running
        
        /path/to/consistent_trees/gen_sussing_forests myconfig.cfg
    
    after the default-format trees have been generated (Section [2.6](#markdown-header-running-the-code), above).
    For convenience, the output halo catalogues are sorted by forest,
    and there is a file called "`sussing_forests.list`" which contains a 
    list of forests, along with the number of halos in each snapshot
    for each forest.
    
    *Note* that the Sussing Merger Trees format uses M200c as the
    default mass.  If this is not the same as the standard mass definition for the trees, but if
    M200c is available in the `EXTRA_PARAMS` fields, then you should 
    specify this in the config file by setting `SUSSING_MASS_FIELD` 
    to the 0-indexed column for M200c in `EXTRA_PARAMS`.  E.g., if
    `EXTRA_PARAM_LABELS` is "`Spin_bullock M200c T/|U|`", then you would
    set `SUSSING_MASS_FIELD = 1` in the config file.
    
3. ### Halo Statistics ###

    In the `OUTBASE` directory, a large number of statistics and logfiles are
    kept about which halos are added and deleted and why.  In particular,
    the files "`statistics_XYZ.list`" and "`cleanup_statistics_XYZ.list`" contain
    information about how many halos are added and deleted as a function
    of halo mass in Phase I and Phase II of the algorithm (respectively).
    To determine how many halos have actually been deleted from the original
    catalogs, subtract the unlinked phantom column (UP) from the deleted halos
    column (D) in the `cleanup_statistics` files.
    
    If too many halos are being deleted, you may want to soften some of the
    consistency requirements (number of timesteps tracked, etc.) or use a
    more consistent halo finder. `;-)`
    
    Information about the accuracy of positions and velocities returned
    by the halo finder is kept in the `metric_statistics` files (for all halos)
    and the `metric_subs_statistics` files (for subhalos).
    
    


CHANGELOG
v1.0:
*) Major update to improve parallelization, especially for the first gravitational consistency stage.
*) Fixed pair of rare bugs that connected unrelated halos together in the same tree; this affected 5 links per million (thanks to Kristin Riebe and Christoph Behrens for finding).  However, the maximum forest size should now be more manageable in many cases.
*) New option to apply fix for Rockstar spins (affecting versions *prior* to Rockstar v0.99.9-RC3); search README for (FIX_ROCKSTAR_SPINS) usage details.
*) Includes conversion code for generating forests in the Sussing Merger Trees format; see README for usage details.
*) New option to limit memory usage (set LIMITED_MEMORY=1 in config file), helpful when dealing with >2048^3 boxes.  Currently, this turns off gzipping outputs (which requires forked processes to remain alive longer), and also limits the parallelization when resorting outputs (which would require more snapshots to be in memory at the same time).
*) Experimental option for binary intermediate output formats, which reduces load/run times for large boxes; search README for INPUT_FORMAT for details.
*) New PDF readme.
*) Added timing info (output in OUTBASE/timing.log).
*) Fixed bug where next_coprogenitor_df_id should have been -1 for the last progenitor at a given redshift.  (Thanks to Morgane Cousin for finding).

v0.99.9.3 RC3 (6/7/2014):
*) Consistent Trees will now ignore all duplicate scale factors in the scale file.  In addition, it will give an error if the scale file is not correctly sorted.
*) Large halo IDs now properly printed in catalog output when using 32-bit versions of perl.
*) Additional accretion rates and accretion times (scale factors) calculated.
*) Now honors non-integer box sizes.
*) Now includes AHF conversion code (beta!) from Yao-Yuan Mao.

v0.99.9.2 RC2 (3/18/2013):
*) Changed PADDING_TIMESTEPS default to 0 to fix a crash in resort_outputs.
*) Updated Universe::Time to fix undefined value warning.

v0.99.9 RC1 (3/13/2013):
*) Additional quantities calculated (half-mass scale factor, accretion rates) in catalog output.
*) Updates to improve handling of a>>1 simulations.

v0.99.5: (Not publicly released).
*) Additional indices (last progenitor ID, next coprogenitor ID) to help with Millennium-format conversion.
*) New file (locations.dat) generated to assist with quickly locating trees in the output files.
*) New file (forests.list) generated to assist with resolving parent-sub links across merger trees.
*) Fixed a crashing problem in rare cases where velocity / position measurements were too precise in the halo finder for the merger tree code to estimate velocity / position errors properly.
*) Fixed an issue whereby duplicate Gadget outputs at the last timestep would cause a crash when generating depth-first IDs.
*) Adjusted last major merger to mean the last scale at which a major merger (satellite) entered the host halo.
*) Switched to favoring the original mvir from the halo catalogs over the smoothed, mass-conserving mass.
*) Fixed a crash due to an off-by-one error (thanks to Steven Rieder for finding!).
*) Fixed an out-of-memory condition which occurred in resort_outputs when reading inputs becomes significantly faster than writing outputs.
*) Fixed a bug whereby parent halos were incorrectly calculated across periodic boundaries for box sizes less than 250 Mpc/h (thanks to Steven Rieder for finding!).

v0.99 (03/2012):
*) Added support for propagating extra halo parameters through the merger trees if your halo finder supports them.
*) Additional columns output to help conversion to Millennium formats: Breadth-first order, Depth-first order, Tree Root ID.
*) Fixed a crashing problem if the last halo catalog did not have all descendants set to -1.

v0.96.2 (12/3/2011):
*) Updated makefiles to fix compilation problems on newer versions of Linux.

v0.96.1 (10/20/2011):
Initial public release


src/halo_io.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <stdint.h>
#include <inttypes.h>
#include <strings.h>
#include <sys/stat.h>
#include <time.h>
#include "gravitational_consistency.h"
#include "masses.h"
#include "check_syscalls.h"
#include "stringparse.h"

extern float min_mvir, max_mvir;
extern float box_size;
FILE *timing_file = NULL;
int64_t timing_start = 0;

void print_timing(char *prog, char *info) {
  char buffer[1024];
  if (!timing_file) {
    snprintf(buffer, 1024, "%s/timing.log", OUTBASE);
    timing_file = check_fopen(buffer, "a");
    timing_start = time(NULL);
    time_t t = timing_start;
    fprintf(timing_file, "%s started at %s", prog, ctime(&t));
    fflush(timing_file);
  }
  if (!info) return;
  fprintf(timing_file, "[%6"PRId64"s] %s\n", (int64_t)time(NULL)-timing_start, info);
  fflush(timing_file);
}

void close_timing_log(void) {
  fprintf(timing_file, "\n");
  fclose(timing_file);
}


void gzip_file(char *filename) {
  char buffer[1024];
  snprintf(buffer, 1024, "gzip -f \"%s\" &", filename);
  if (!LIMITED_MEMORY)
    system(buffer);
}

void print_halo(FILE *o, struct tree_halo *th) {
  int64_t mmp, i;
  int64_t flags;
  if (!th) {
    if (!strcasecmp(INPUT_FORMAT, "BINARY")) return;
    fprintf(o, "#ID DescID Mvir Vmax Vrms Rvir Rs Np X Y Z VX VY VZ Jx Jy Jz spin %s Phantom MMP Suspicious? PID UPID Tracked Tracked_Single_MMP Num_MMP_Phantoms Original_ID Last_mm\n", EXTRA_PARAM_LABELS);
     fprintf(o, "#Units: Masses in Msun / h\n"
            "#Units: Positions in Mpc / h (comoving)\n"
            "#Units: Velocities in km / s (physical)\n"
            "#Units: Angular Momenta in (Msun/h) * (Mpc/h) * km/s (physical)\n"
            "#Units: Radii in kpc / h (comoving)\n");
    return;
  }
  if (!strcasecmp(INPUT_FORMAT, "BINARY")) {
    fwrite(th, sizeof(struct tree_halo), 1, o);
    return;
 }
  mmp = (th->flags & MMP_FLAG) ? 1 : 0;
  flags = 0;
  if (th->flags & SUSPICIOUS_LINK_FLAG) flags = 1;
  if (th->flags & UNPHYSICAL_LINK_FLAG) flags = 2;
  if (th->flags & MERGER_FLUCTUATION_FLAG) flags = 3;
  flags += (th->flags & (TOO_MANY_PHANT_FLAG | SHORT_TRACK_FLAG | MISTRACKED_SUB_FLAG));
  fprintf(o, "%"PRId64" %"PRId64" %.3e %.2f %.2f %.3f %.3f %"PRId64" %.5f %.5f %.5f %.2f %.2f %.2f %.3e %.3e %.3e %.5f",
	  th->id, th->descid, th->mvir, th->vmax, th->vrms, th->rvir, th->rs,
	  th->np, th->pos[0], th->pos[1], th->pos[2], th->vel[0], th->vel[1],
	  th->vel[2], th->J[0], th->J[1], th->J[2], th->spin);
  for (i=0; i<EXTRA_PARAMS; i++) {
    if (th->extra_params[i] == ((double)((int64_t)th->extra_params[i])))
      fprintf(o, " %"PRId64, (int64_t)th->extra_params[i]);
    else
      fprintf(o, " %e", th->extra_params[i]);
  }
  fprintf(o, " %"PRId64" %"PRId64" %"PRId64" %"PRId64" %"PRId64" %"PRId64" %"PRId64" %"PRId64" %"PRId64" %f\n",
	  th->phantom, mmp, flags, th->pid, th->upid, th->tracked,
	  th->tracked_single_mmp, th->num_mmp_phantoms, th->orig_id, th->last_mm);
}

void common_print_halos(char *filename, struct tree_halo *h, int64_t num_halos, int64_t gzip_halos)
{
  int64_t i;
  FILE *o;

  o = check_fopen(filename, "w");
  if (!strcasecmp(INPUT_FORMAT, "BINARY")) {
    fwrite(h, sizeof(struct tree_halo), num_halos, o);
  } else {
    print_halo(o, NULL);
    for (i=0; i<num_halos; i++) print_halo(o, &(h[i]));
  }
  fclose(o);
  if (gzip_halos) gzip_file(filename);
}

int really_beyond_mmp_ratio(struct tree_halo h1, struct tree_halo h2) {
  float m1, m2, v1, v2;
  if (h1.mvir > h2.mvir) { m1 = h1.mvir; m2 = h2.mvir; }
  else { m1 = h2.mvir; m2 = h1.mvir; }
  if (m1*0.3*MIN_MMP_MASS_RATIO > m2) return 1;

  if (h1.vmax > h2.vmax) { v1 = h1.vmax; v2 = h2.vmax; }
  else { v2 = h1.vmax; v1 = h2.vmax; }
  if (v1*0.5*MIN_MMP_VMAX_RATIO > v2) return 1;

  return 0;
}

int beyond_mmp_ratio(struct tree_halo h1, struct tree_halo h2) {
  float m1, m2, v1, v2;
  if (h1.mvir > h2.mvir) { m1 = h1.mvir; m2 = h2.mvir; }
  else { m1 = h2.mvir; m2 = h1.mvir; }
  if (m1*MIN_MMP_MASS_RATIO > m2) return 1;

  if (h1.vmax > h2.vmax) { v1 = h1.vmax; v2 = h2.vmax; }
  else { v2 = h1.vmax; v1 = h2.vmax; }
  if (v1*MIN_MMP_VMAX_RATIO > v2) return 1;

  return 0;
}

void build_id_conv_list(struct halo_stash *h)
{
  int64_t n, ids;
  if (h->num_halos < 1) return;

  h->min_id = h->max_id = h->halos[0].id;
  for (n=0; n<h->num_halos; n++) {
    if (h->max_id < h->halos[n].id) h->max_id = h->halos[n].id;
    if (h->min_id > h->halos[n].id) h->min_id = h->halos[n].id;
  }

  ids = (1 + h->max_id - h->min_id);
  h->id_conv = (int64_t *)check_realloc(h->id_conv, sizeof(int64_t)*ids, "halo index conversion");
  for (n=0; n<ids; n++) h->id_conv[n]=-1;

  for (n=0; n<h->num_halos; n++)
    h->id_conv[h->halos[n].id - h->min_id] = n;

  assert(h->min_id >= 0); //Otherwise, logic problems occur in rest of code.
}


void add_halo(struct halo_stash *h, struct tree_halo d)
{
  if (((h->num_halos)%1000)==0) {
    int64_t array_size = h->num_halos+1000;
    h->halos = (struct tree_halo *)
      check_realloc(h->halos,sizeof(struct tree_halo)*array_size, "Adding new halo.");
  }

  h->halos[h->num_halos] = d;
  h->num_halos++;
}

void load_halos(char *filename, struct halo_stash *h, float scale, int dead)
{
  FILE *input;
  char buffer[1024];
  struct tree_halo d = {0};
  int64_t n;
  int64_t mmp, flags, phantom;
  float delta_mvir = delta_vir(scale);
  float mean_density = 2.77519737e11*Om; //(Msun/h) / (comoving Mpc/h)^3
  float vir_density = delta_mvir*mean_density;
  int64_t expected_inputs = 28+EXTRA_PARAMS;
  int64_t regular_inputs = 18;
  SHORT_PARSETYPE;
#define NUM_INPUTS (28+MAX_EXTRA_PARAMS)
  enum short_parsetype stypes[NUM_INPUTS] = 
    { D64, D64, F, F, F, F, F, D64, F, F, F, F, F, F, F, F, F, F, D64, D64, D64, D64, D64, D64, D64, D64, D64, F };
  enum parsetype types[NUM_INPUTS];
  void *data[NUM_INPUTS] = {&(d.id),
			    &(d.descid), &(d.mvir), &(d.vmax), &(d.vrms),
			    &(d.rvir), &(d.rs), &(d.np), &(d.pos[0]),
			    &(d.pos[1]), &(d.pos[2]), &(d.vel[0]), 
			    &(d.vel[1]), &(d.vel[2]), &(d.J[0]), &(d.J[1]),
			    &(d.J[2]), &(d.spin),
			    &(phantom), &(mmp), &(flags), &(d.pid), &(d.upid),
			    &(d.tracked), &(d.tracked_single_mmp), 
			    &(d.num_mmp_phantoms), &(d.orig_id), &(d.last_mm)};

  int64_t read_ascii = 1;
  if (!strcasecmp(INPUT_FORMAT, "BINARY")) read_ascii = 0;

  gen_ff_cache();
  d.num_prog = d.flags = d.phantom = 0;
  d.mmp_id = -1;
  for (n=0; n<NUM_INPUTS; n++) types[n] = stypes[n];
  for (n=expected_inputs-1; n>=regular_inputs+EXTRA_PARAMS; n--) {
    types[n] = types[n-EXTRA_PARAMS];
    data[n] = data[n-EXTRA_PARAMS];
  }
  for (n=regular_inputs; n<regular_inputs+EXTRA_PARAMS; n++) {
    types[n] = PARSE_FLOAT64;
    data[n] = &(d.extra_params[n-regular_inputs]);
  }
  
  int64_t old_halos = h->num_halos;
  input = check_fopen(filename, "r");
  if (read_ascii) {
    while (fgets(buffer, 1024, input)) {
      if (buffer[0] == '#') continue;
      mmp = flags = 0;
      n = stringparse(buffer, data, (enum parsetype *)types, expected_inputs);
    
      if (n < regular_inputs-4) continue;
      if (n < regular_inputs) { d.J[0] = d.J[1] = d.J[2] = d.spin = 0; }
      if (n < regular_inputs+EXTRA_PARAMS+1) d.phantom = 0;
      else d.phantom = phantom;
      if (n < regular_inputs+EXTRA_PARAMS+4) d.pid = -1;
      if (n < regular_inputs+EXTRA_PARAMS+5) d.upid = -1;
      if (n < regular_inputs+EXTRA_PARAMS+6) d.tracked = 0;
      if (n < regular_inputs+EXTRA_PARAMS+7) d.tracked_single_mmp = 0;
      if (n < regular_inputs+EXTRA_PARAMS+8) d.num_mmp_phantoms = 0;
      if (n < regular_inputs+EXTRA_PARAMS+9) d.orig_id = -1;
      if (n < regular_inputs+EXTRA_PARAMS+10) d.last_mm = 0;

      d.mvir = fabs(d.mvir);
      if (!(d.mvir>0)) continue;
      if (!(d.rvir > 0))
	d.rvir = cbrt(d.mvir / (4.0*M_PI*vir_density/3.0)) * 1000.0;
      if (!(d.rs > 0)) d.rs = d.rvir / concentration(d.mvir, scale);

      d.flags = 0;
      if (mmp) d.flags |= MMP_FLAG;
      if ((flags & 3)==1) d.flags |= SUSPICIOUS_LINK_FLAG;
      if ((flags & 3)==2) d.flags |= UNPHYSICAL_LINK_FLAG;
      if (dead) d.flags |= DEAD_HALO_FLAG;
      add_halo(h, d);
    }
  }
  else {
    struct stat s;
    assert(!fstat(fileno(input), &s));
    int64_t num_inputs = s.st_size / (sizeof(struct tree_halo));
    if (num_inputs * sizeof(struct tree_halo) != s.st_size) {
      fprintf(stderr, "Error: file format in %s not recognized as %s!\n", filename, INPUT_FORMAT);
      exit(1);
    }
    int64_t alloc_size = num_inputs+1000;    
    h->halos = (struct tree_halo *)
      check_realloc(h->halos,sizeof(struct tree_halo)*(old_halos+alloc_size), "Adding new halos.");
    fread(h->halos+old_halos, sizeof(struct tree_halo), num_inputs, input);
    h->num_halos += num_inputs;
  }
  fclose(input);

  for (int64_t i=old_halos; i<h->num_halos; i++) {
    mmp = h->halos[i].flags & MMP_FLAG;
    flags = 0;
    if (h->halos[i].flags & SUSPICIOUS_LINK_FLAG) flags |= 1;
    if (h->halos[i].flags & UNPHYSICAL_LINK_FLAG) flags |= 2;
    h->halos[i].flags = 0;
    if (mmp) h->halos[i].flags |= MMP_FLAG;
    if ((flags & 3)==1) h->halos[i].flags |= SUSPICIOUS_LINK_FLAG;
    if ((flags & 3)==2) h->halos[i].flags |= UNPHYSICAL_LINK_FLAG;
    if (dead) h->halos[i].flags |= DEAD_HALO_FLAG;
    if (min_mvir==0) min_mvir = h->halos[i].mvir;
    if (h->halos[i].mvir < min_mvir) min_mvir = h->halos[i].mvir;
    if (h->halos[i].mvir > max_mvir) max_mvir = h->halos[i].mvir;
    for (n=0; n<3; n++) if (h->halos[i].pos[n] > box_size) box_size = h->halos[i].pos[n];
  }

  if (!BOX_WIDTH)
    box_size = (int) (box_size + 0.5);
  else
    box_size = BOX_WIDTH;
}


void read_outputs(float **output_scales, int64_t **outputs, int64_t *num_outputs) {
  FILE *input;
  int64_t n;
  char buffer[1024];
  
  input = check_fopen(SCALEFILE, "r");
  while (fgets(buffer, 1024, input)) {
    if (((*num_outputs)%1000 == 0)) {
      *output_scales = (float *)check_realloc(*output_scales, sizeof(float)*((*num_outputs)+1000), "output scales");
      *outputs = (int64_t *)check_realloc(*outputs, sizeof(int64_t)*((*num_outputs)+1000), "output numbers");
    }
    n = sscanf(buffer, "%"SCNd64" %f", &((*outputs)[*num_outputs]), &((*output_scales)[*num_outputs]));
    if (n<2) continue;
    if (num_outputs[0]) {
      float scale_now = output_scales[0][num_outputs[0]];
      float last_scale = output_scales[0][num_outputs[0]-1];
      if (scale_now == last_scale) continue;
      if (scale_now < last_scale) {
	fprintf(stderr, "[Error] Scale file (%s) not sorted in ascending order!\n", SCALEFILE);
	exit(1);
      }
    }
    *num_outputs = (*num_outputs) + 1;
  }
  fclose(input);

  *output_scales = (float *)check_realloc(*output_scales, sizeof(float)*((*num_outputs)), "output scales");
  *outputs = (int64_t *)check_realloc(*outputs, sizeof(int64_t)*((*num_outputs)), "output numbers");
}
